### PR TITLE
Enable support for tornado 6.2

### LINKED ISSUE #6669: Tornado 6.2 compatibility
Tornado 6.2 was just released and we're hitting a few errors on CI. So far, the only thing I observed was a deprecation warning that is converted to an error. We should restore compatibility and remove the pinning introduced in https://github.com/dask/distributed/pull/6668

cc @graingert

### PR DESCRIPTION
**Pull request by wence-**
Most of the work in deprecation was done as part of #6680; with
tornado 6.2 in the few tests that still rely on creating an event loop
we must catch deprecation warnings from tornado in addition to those
from asyncio.

I suppose that the next step would be to remove support in
`LoopRunner` and friends for this deprecated functionality, and then
cull the tests that are checking for this behaviour, but I have not
done this here.

This ports over the version pin changes from #6991.

Closes #6669.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/deploy/tests/test_adaptive.py [hunk 1]
```diff
@@ -347,7 +347,7 @@ def key(ws):
         adaptive = cluster.adapt(minimum=1)
         await adaptive.adapt()
 
-        while len(cluster.scheduler.workers) == 4:
+        while len(cluster.scheduler.workers) >= 3:
```
Comment (wence-) [inline distributed/deploy/tests/test_adaptive.py:351]: This is unrelated to the tornado changes, but this test was intermittently failing locally.

We're attempting to scale down from four to two workers (based on the grouping key, IIUC), but we might not have torn down both additional workers by the time this loop exits if just checking that we no longer have four workers.

File: continuous_integration/recipes/distributed/meta.yaml [hunk 2]
```diff
@@ -42,7 +42,7 @@ requirements:
     - sortedcontainers !=2.0.0,!=2.0.1
     - tblib >=1.6.0
     - toolz >=0.8.2
-    - tornado >=6.0.3,<6.2
+    - tornado >=6.0.3
```
Comment (wence-) [inline continuous_integration/recipes/distributed/meta.yaml:45]: Note that I just copied over the changes for #6991; in 3.10 pinning exactly to 6.2 ensures that we actually test with 6.2 in one CI setting. I don't know if the looser pins should all explicitly request `>= 6.0.3` (at present they pick up 6.1)

**Comment by wence-** [comment]
Windows fail appears to be flaky/bad test #7135;

File: distributed/deploy/tests/test_spec_cluster.py [hunk 3]
```diff
@@ -83,6 +83,7 @@ def test_spec_sync(loop):
 
 
 @pytest.mark.filterwarnings("ignore:There is no current event loop:DeprecationWarning")
+@pytest.mark.filterwarnings("ignore:make_current is deprecated:DeprecationWarning")
 def test_loop_started_in_constructor(cleanup):
     # test that SpecCluster.__init__ starts a loop in another thread
     cluster = SpecCluster(worker_spec, scheduler=scheduler, loop=None)
```
Comment (wence-) [inline distributed/deploy/tests/test_spec_cluster.py:89]: AIUI, this is explicitly deprecated functionality (building a `LoopRunner` without an active loop), but unpicking that seems like a larger job.

**Comment by wence-** [comment]
Since I didn't manage to get this into 2022.10.2, I'd like to have a more comprehensive go this time.

As it stands, this PR just kicks the can down the road wrt deprecated functionality: distributed internally doesn't use the deprecated functionality so I only needed to touch tests.

Depending on whether we feel the deprecation cycle has been long enough (#6680 came in in July) the larger set of changes would be to _remove_ all the deprecated functionality in `LoopRunner` and so forth, and then pull out the tests of this functionality completely.

Before I head in this direction, do we want that now?

**Comment by fjetter** [comment]
I still see `IOLoop.make_current` being used in `Nanny._run`. This might be the final invocation in our actual code base.

If this is just about ripping out all deprecated calls concerning the loop runner, I'm OK with this. afaik, most downstream consumers were migrated. @jacobtomlinson I believe you were involved in these efforts.

File: continuous_integration/environment-3.10.yaml [hunk 4]
```diff
@@ -38,7 +38,7 @@ dependencies:
   - sortedcollections
   - tblib
   - toolz
-  - tornado<6.2
+  - tornado=6.2
```
Comment (graingert) [inline continuous_integration/environment-3.10.yaml:41]:
```suggestion
  - tornado>=6.2
```

**Review by graingert — CHANGES_REQUESTED** [review]

File: continuous_integration/environment-3.10.yaml (hunk 4 above)
Comment (wence-) [reply continuous_integration/environment-3.10.yaml:41]: This came from #6991, should we update these pins for _all_ environments? (not just 3.10)

Comment (graingert) [reply continuous_integration/environment-3.10.yaml:41]: yes probably all the environments need it, otherwise it's fighting the pin from distributed==2022.10.2

https://github.com/dask/distributed/pull/6991#pullrequestreview-1095039021

File: distributed/tests/test_client.py [hunk 5]
```diff
@@ -2206,12 +2205,13 @@ async def test_multi_client(s, a, b):
...
 
 def long_running_client_connection(address):
-    with pristine_loop():
+    async def run():
         c = Client(address)
         x = c.submit(lambda x: x + 1, 10)
-        x.result()
         sleep(100)
 
+    asyncio.new_event_loop().run_until_complete(run())
```
Comment (graingert) [inline distributed/tests/test_client.py:2213]:
```suggestion
    asyncio.run(run())
```

File: continuous_integration/environment-3.10.yaml (hunk 4 above)
Comment (wence-) [reply continuous_integration/environment-3.10.yaml]: Done, hopefully I caught all the spots.

File: distributed/tests/test_client.py [hunk 6]
```diff
@@ -5538,23 +5538,18 @@ async def test_future_auto_inform(c, s, a, b):
             await asyncio.sleep(0.01)
 
 
-@pytest.mark.filterwarnings("ignore:There is no current event loop:DeprecationWarning")
-@pytest.mark.filterwarnings("ignore:make_current is deprecated:DeprecationWarning")
-@pytest.mark.filterwarnings("ignore:clear_current is deprecated:DeprecationWarning")
 def test_client_async_before_loop_starts(cleanup):
```
Comment (graingert) [inline distributed/tests/test_client.py:5541]: the name of the test doesn't match what's done in the body of the test anymore

File: distributed/tests/test_client.py [hunk 7]
```diff
@@ -2206,12 +2205,13 @@ async def test_multi_client(s, a, b):
 
 
 def long_running_client_connection(address):
-    with pristine_loop():
+    async def run():
         c = Client(address)
         x = c.submit(lambda x: x + 1, 10)
```
Comment (graingert) [inline distributed/tests/test_client.py:2210]:
```suggestion
        with Client(address) as c:
            x = c.submit(lambda x: x + 1, 10)
            sleep(100)
```

File: distributed/tests/test_client.py [hunk 8]
```diff
@@ -2206,12 +2205,13 @@ async def test_multi_client(s, a, b):
...
-        x.result()
         sleep(100)
```
Comment (graingert) [inline distributed/tests/test_client.py:2211]:
```suggestion
```

File: distributed/deploy/tests/test_spec_cluster.py (hunk 3 above)
Comment (graingert) [reply distributed/deploy/tests/test_spec_cluster.py:89]: building a LoopRunner without an active loop is not deprecated - LoopRunner will use `asyncio.run` to start a new loop in the future

Comment (wence-) [reply distributed/deploy/tests/test_spec_cluster.py:89]: Perhaps I misunderstand the current deprecation warnings then:

https://github.com/dask/distributed/blob/main/distributed/utils.py#L435-L457

Right now, if we don't pass a `loop` object, and ask for an asynchronous `LoopRunner` and there is not asyncio loop running we get a deprecation warning. So presumably this branch should disappear (if `asynchronous=True` one must pass a loop object).

If no loop is passed and we're not asynchronous, then the idea is that we get a new `IOLoop` for a thread (eventually). In this case, I suppose we must build a new loop with `asyncio.new_event_loop()`? AIUI, `asyncio.run` will build a new loop, run a coroutine, and then tear the loop down, which doesn't _seem_ like what is desired here.

Contrarily, if a `loop` _is_ provided and the wrapped asyncio event loop is not running, we get a deprecation warning. So presumably this should become unsupported behaviour now.

**Comment by graingert** [comment]
> I still see `IOLoop.make_current` being used in `Nanny._run`. This might be the final invocation in our actual code base.

I opened https://github.com/dask/distributed/pull/7240 to remove the calls from `Nanny._run`

File: distributed/deploy/tests/test_spec_cluster.py (hunk 3 above)
Comment (graingert) [reply distributed/deploy/tests/test_spec_cluster.py:89]: > In this case, I suppose we must build a new loop with asyncio.new_event_loop()? AIUI, asyncio.run will build a new loop, run a coroutine, and then tear the loop down, which doesn't seem like what is desired here.

yes using asyncio.run in this way is the plan, the change in `LoopRunner._start_unlocked.run_loop` will look like:

```python
        def run_loop(loop=self._loop):
            async def _():
                self._stop_event = asyncio.Event()
                self._loop = IOLoop.current()
                loop_cb()
                await self._stop_event.wait()

            try:
                asyncio.run(_())
            except Exception as e:
                start_exc[0] = e
            finally:
                done_evt.set()
```

File: requirements.txt [hunk 9]
```diff
@@ -9,7 +9,7 @@ psutil >= 5.0
 sortedcontainers !=2.0.0, !=2.0.1
 tblib >= 1.6.0
 toolz >= 0.8.2
-tornado >= 6.0.3, <6.2
+tornado >= 6.2
```
Comment (ivanov) [inline requirements.txt:12]: tiny drive-by comment, is the requirement to upgrade to `>=6.2` necessary here? I ask because I saw that in https://github.com/dask/distributed/pull/7165 there was code backported for when tornado < 6.2.0

Comment (wence-) [reply requirements.txt:12]: It will not be, but I want to ensure that this tests against 6.2 or later in CI for now.

Comment (fjetter) [reply requirements.txt:12]: requirements.txt changes will translate directly to the packaged version. If you only want to force CI to use this, please only pin in `continuous_interation/environment*.yml`

**Review by graingert — APPROVED** [review]

**Comment by wence-** [comment]
Let me fix the lint errors, but I am pretty sure this is not yet ready unfortunately.

**Comment by wence-** [comment]
FWIW, I've had a go at trying to understand how `LoopRunner`s interact with the rest of the code and I'm overly confused about what the semantics should be. So if someone with a better understanding than me knows what to do here, I would encourage them to step in. The more recent changes turn the deprecation warnings in `LoopRunner` into runtimeerrors, so we can see where things break in the test suite.

**Comment by fjetter** [comment]
closed and incorporated by https://github.com/dask/distributed/pull/7286
